frb_codegen/src/generator/rust/ty.rs

```diff
@@ -64,10 +56,47 @@ pub trait TypeRustGeneratorTrait {
     fn imports(&self) -> Option<String> {
         None
     }
+
+    fn get_context(&self) -> &TypeGeneratorContext;
+
+    fn get_shared_mod_name_if_type_shared(&self, ty: &IrType) -> Option<String> {
+        if self.get_context().ir_file.is_type_shared_by_safe_ident(ty) {
+            return SHARED_MODULE.with(|data| {
+                let cloned = data.borrow().clone();
+                if cloned.is_none() {
+                    panic!("in instance in charge of `{}`: checking shared for type \"{:?}\", it is shared indeed, but the shared module name is None",
+                        self.get_context().type_name, ty
+                    );
+                }
+                cloned
+            });
+        }
+        None
+    }
+
+    fn get_shared_module_of_a_type(&self, ir_type: &IrType) -> Option<String> {
+        match ir_type {
+            IrType::Optional(inner_type) => {
+                self.get_shared_mod_name_if_type_shared(&inner_type.inner)
+            }
+            _ => self.get_shared_mod_name_if_type_shared(ir_type),
+        }
+    }
+
+    fn get_wire2api_prefix(&self, ir_type: &IrType) -> String {
```

Comment:
after all the review, I realize there are a few utility functions dealing with "change the generated code a bit according to shared or non-shared". 

Thus, I propose that, we should extract **all these** utility functions (including the ones that we have already written, and the ones I propose in my comments above), into a separate utility struct (or file or mod or whatever). for example:

```rust
struct UtilityForGeneratingCodeRelatedToShared;
impl UtilityForGeneratingCodeRelatedToShared {
  fn wire2api_prefix() { ... the function we are discussing here }
  fn shared_impl_related_thing() { ... IIRC I proposed something related to generating _sharedImpl or whatever}
  fn yet_another_one() { ... I remember there is at least another utility related to us ... }
}
```

by doing these, we are super clear what utility functions we have and happily use them